### ISSUE TITLE
Improve remote mining

### ISSUE THREAD

**Issue by SirLovi** [ScreepsOCS/screeps.behaviour-action-pattern#206]
Having dedicated miners right next to spawns with containers in exploited rooms and then having remote-haulers to pick up the energy could be more energy efficient.

![capture](https://cloud.githubusercontent.com/assets/20193391/19236472/efbf286e-8ef7-11e6-8488-02739481e353.PNG)

**Comment by cyberblast** [comment]
Sure, why not... 

* reserving is working already (white/gray flag). 
* miners + haulers for exploitation rooms 
  * correlates with the "Operator"  idea (merge workers with privateers & pioneers) 
  * miners are probably too slow, may need some additional MOVE parts eventually 
  * may be worth a try... 

I'd propose to send 1 miner and 1 hauler + 1 privateer (as hauler and mechanic) per exploit flag. 
Current best practise is to put 1 flag per room, would be 1 per source then. 

On the other hand... remote harvesting rules are about to be changed drastically soon (no more boost by reserving). Maybe it will be better to wait until then? I'm unsure how to face it...

**Comment by erik-smit** [comment]
I took an ugly stab at this at https://github.com/erik-smit/screeps.behaviour-action-pattern/commit/0f46214936f3fff86c869e53cdc2e6a1768e4312

I turned the privateer in a dedicated miner and added a 'haulateer' for the hauling.

Issues I ran into:

  * privateer cloaks flag when source is empty, preventing flagDir from finding it for the haulateer
  * privateer also does the building and maintaining in the exploited rooms. With current code it still does, until the container is built. Who will do it when there is a dedicated miner?

**Comment by cyberblast** [comment]
Maybe we could do it like we did with owned rooms. 
In the beginning there are only worker. Then we add miner and hauler, keeping one worker. 

The alternative could be to add one work part to the hauler and repair on drive-by. 
But construction would be quite slow with that. But we could use pioneers for that. or simply a worker. Those types could be merged anyway (worker, privateer, pioneer).

Maybe we can also use the current hauler setup, adding a new action to get the energy from remote rooms. Similar for miner. 

I'd propose to wait until we have the spawnQueue. I'm currently working on that. Will make it possible to also set a destiny for the new creep. This can be evaluated one the creep finished spawning (e.g. for the miner to assign it to a room).

**Comment by Dasku** [comment]
This is definetly something that's been on my mind also. I think remote miners (with a container) + haulers should give an efficiency increase. 10k per hour per source seems like a pretty good target. Home rooms harvest 20k per hour pretty easily, however I sometimes struggle to get 20k per hour from Privateers (even if they have 2-3 rooms available, so 4-6 sources).

This same concept would also work perfectly once we can mine centre rooms. Miners (bigger as centre room sources have 4k energy), and haulers with a SK guard to keep them safe. 

As you say, solving the road repair issue would be a problem if haulers dont have WORK. A small worker sent to repair in the room could be an easy fix? Or a bigger worker/privateer which takes turns going to each room within range to carry out repairs before moving on to next room (could be a lot more complicated). Or add 1-2 WORK to haulers, but then you sacrifice CARRY space.

**Comment by erik-smit** [comment]
> I'd propose to wait until we have the spawnQueue. I'm currently working on that. Will make it possible to 
> also set a destiny for the new creep. This can be evaluated one the creep finished spawning (e.g. for the 
> miner to assign it to a room).

I'd like to take another stab at this.

> I'd propose to send 1 miner and 1 hauler + 1 privateer (as hauler and mechanic) per exploit flag. 
> Current best practise is to put 1 flag per room, would be 1 per source then.

So, I would...

  * modify the hauler and miner maxCount to include remote mining 
  * modify hauler and miner spawning to include destiny
  * modify hauler and miner behaviour to work towards 'destiny' instead of the current independent decision making?
  * stop the privateers from mining, and just focussing on building, repairing?

does this make sense?

I don't really have an idea yet where the destinies would be introduced.

Do I understand correctly that the spawnQueue isn't really used yet?

**Comment by cyberblast** [comment]
> Do I understand correctly that the spawnQueue isn't really used yet? 

Confirmed. It's implemented and I tested it adding something manually. But nothing is really using it yet. 

> modify the hauler and miner maxCount to include remote mining

hmm no. the spawnQueue is not respecting these properties. The idea is an event driven approach, where the task orders the required creep.  
Currently we are simply iterating spawns and check if it is allowed to spawn something. This is done using rules defined in the setup files. 

The spawnQueue will simply spawn what is ordered. So there needs to be an "event" like "exploit flag found without assigned harvester" then order that harvester and register it on the flag. Something like that.

> modify hauler and miner spawning to include destiny 

yes, we should redesign the current spawning logic, also for owned rooms. 

> modify hauler and miner behaviour to work towards 'destiny' instead of the current independent decision making? 

the hauler will still have to make decisions. aka checking an action priority array. But we could assign it to a room... 
the miner doesn't make any decisions anyway. but it will also have to be assigned to a room using the "destiny" flag

> stop the privateers from mining, and just focussing on building, repairing? 

I'm unsure about that. Maybe i could make sense to let them decide it. Check if there is a source without a dedicated miner. if so it may go harvesting. 

> I don't really have an idea yet where the destinies would be introduced. 

its a property for the object you can put into the spawn queue. 
sample: `Game.rooms['W44N3'].spawnQueueHigh.push({parts:[MOVE],name:'max',setup:'Healer',cost:50,destiny:'W44N3'})`

**Comment by erik-smit** [comment]
> The spawnQueue will simply spawn what is ordered. So there needs to be an "event" like "exploit flag 
> found without assigned harvester" then order that harvester and register it on the flag. Something like that.

How would that event be fired?

I imagine something iterating over something every now and then.
Just like the SPAWN_INTERVAL we do now.

Or do you imagine something like a creep who currently targets Flag123 dieing, and firing the event letting know the system Flag123 is now unassigned?

**Comment by cyberblast** [comment]
we are currently iterating creeps and register them to actions and targets as stored in their memory in the beginning of each loop. 
after that we could iterate each flag and see how it is. if there is enough registered on it. 

hmm - we are already iterating flags, but before creeps. I'm unsure if there is a reason for this, will need to have a closer look. 
For a first approach we could simply iterate flags a second time. 

yes, a simple first step could be migrating claimer and combat creeps to the new system. These are also most urgent, as their current calculation system is not working correctly.

**Comment by erik-smit** [comment]
Currently it iterates the flags to
  * push them onto FlagDir.list
  * decrement cloaking
  * delete targetOf

I think it runs before the Creep-loop and deletes targetOf, so the Creep-loop can take the appropriate behaviour.

Are you sure looping over the flags again in FlagDir to determine which Creeps to spawnQueue would be the way to go?

Maybe each creep role could export a function that contains the role-specific spawning behaviour?
Maybe a creep.setup.role.loop function per role

**Comment by cyberblast** [comment]
thats more or less how the current system works. we got a setup check per role.

**Comment by erik-smit** [comment]
Currently there's a spawn-loop in spawn.js, which loops over all roles and runs isValidSetup from creep.setup.js, which checks parameters exported by individual roles.

Now we want something that queues a spawn based on... untargeted flags? 

I'm not really sure where to start here.

Just looping over all flags, checking if they don't have a targetOf, then checking which color they are and choose a role based on that sounds wrong to me.
And you might spawn want to spawn multiple things based on one flag. For instance a miner and hauler for one exploit flag.

**Comment by cyberblast** [comment]
But that's exactly the reason to spawn a claimer and some privateers (to stay at the current remote harvesting system for the moment). 

There is an exploit flag. And targetOf does not only count creeps, but also contains the required data about which creep it is.

**Comment by cyberblast** [comment]
1. We should create a feature branch for this. 
2. disable "classical" spawning for (i.e.) claimers
3. iterate flags
4. define flag analysis & spawning events (per flag type). room selection (maybe in room.js). creep body definition (this last point could still be defined using setup files).

**Comment by erik-smit** [comment]
I'm thinking about this:

```
        Spawn.prototype.flagloop = function(){
            let room = this.room;
            var probe = flag => {
                if (flag.targetOf === undefined || !_.some(flag.targetOf, {'creepType': 'claimer'}) {
                    this.room.spawnQueueLow.push(blah);
                }
            };
            _.forEach(Game.flags, probe);            
        } 
```

But this is probably not right, sensible, etc.

> define flag analysis & spawning events (per flag type). room selection (maybe in room.js). creep body definition (this last point could still be defined using setup files).

Maybe the flag analysis, spawning events, etc. could be defined using the setup-files?

Either have a flap-loop deferring to a Creep.setup.role.probe() function, or maybe have a flag-loop inside every Creep.setup.role.probe() function.

**Comment by cyberblast** [comment]
> But this is probably not right. 

Flag iteration should not be for each spawn. But only once globally. 

> also, would it make more sense to paste the code in here to discuss while building, or to commit to the branch even though my code is still shit/broken/etc.? 

I have no problem with broken feature branches. It makes it easier to share the code you are currently working on and discuss it. 

> Also, do I understand correctly that the spawnQueueHigh/Low variables are variables that are cleared every loop? 

No, it goes to/comes from the persistent memory. 

> Maybe the flag analysis, spawning events, etc. could be defined using the setup-files?

No, I don't think it makes sense to define that for every creep type again, e.g. what should be probed for an exploit flag. 
flag analysis & spawning events => per flag type
room selection (for spawning the creep) => I'm unsure, but it could make sense to be implemented in room.js
creep body definition => setup files

> Either have a flap-loop deferring to a Creep.setup.role.probe() function, or maybe have a flag-loop inside every Creep.setup.role.probe() function. 

I'd prefer to have 1 flag loop (or maybe a second one as addition to the one we have now). I don't think it is required to probe for the same exploit flag multiple time per loop. there is 1 analysis if the flags required crowd is saturated. if not then look for the best room for spawning and order whats required.

**Comment by erik-smit** [comment]
> Flag iteration should not be for each spawn. But only once globally.

Ahh, I missed that Spawn.prototype.loop gets called for every room by Spawn.loop.

So where would we do the Flag iteration for spawning?

> flag analysis & spawning events => per flag type

FlagDir.invade.exploit.js
FlagDir.claim.reserve.js
?

Or like a switch/case in the flag iteration loop?

Or like a bunch of:
`FlagDir.filter(FLAG_COLOR.invade.exploit).foreach(probe)
FlagDir.filter(FLAG_COLOR.claim.reserve).foreach(probe)`?

**Comment by cyberblast** [comment]
All good approaches, but I think the best would indeed be 1 iteration with maybe switches or ifs (I don't like switch, but ok), even if its not very "beautiful", but it would be simple and I believe the most performant. 

hmm we could maybe put the conditions, which are probed against back into the setup files. 

something like 

```
//pseudo
flags = [...]
creepTypes = [...]

foreach(flag in flags){
   foreach(type in creepTypes){
      creep = type.requiredCreep(flag);
      if( creep ){
         Room.bestSpawnRoomFor(flag).spawnQueue.push(creep);
      }
   }
}

```

**Comment by erik-smit** [comment]
I was trying to implement requiredCreep, and I ran into the issue of what it is going to return.
The parts-layout of a creep is dependant on the rcl of the room.
So, type.requiredCreep can't return the parts-layout, or it will need to know the room it'll be spawned from.

**Comment by erik-smit** [comment]
I realized that I can test my code in simulation, but only for one room, so I can't test claimer there, but I could test something like 'worker'.

I'm not sure how to implement it there.

workers are spawned because we want at least 1 in a room where we have a spawn.
So we'd also need a foreach spawn-loop? or foreach room-where-we-have-a-spawn-loop? 

I imagine something this in each creep.setup.type.js
```
setup.spawnLoop = function() {
  foreach(room in rooms) {
    if (we need more workers here than we currently have) {
      room.spawnQueue.push(parts[room.rcl]);
    }
  }
}
```

**Comment by cyberblast** [comment]
> I was trying to implement requiredCreep, and I ran into the issue of what it is going to return.
The parts-layout of a creep is dependant on the rcl of the room.
So, type.requiredCreep can't return the parts-layout, or it will need to know the room it'll be spawned from.

hmm, how about something like a task list. ok this is now going even beyond the current idea, but would be the the final idea anyway (there is also an issue for that somewhere ("turn task assignment to task requestment")). 
A flag is a task (e.g. exploit). 
An owned room is a big task or many smaller, tbd 

That would also make it easier to complete leave flags somewhen and only/also iterate adjacent rooms, to automatically exploit them. 

```
Task: {
   Room room,
   bool requiresCreep(), 
   creep queueCreep(room)
}
```

This would require to define the possible tasks, maybe as own modules... 

```
task.exploit.js: 
{
  requiresCreep: () => { 
    return check required creeps... you get the idea...
  }; 
[...]
}
```

we could then iterate flags, rooms etc. creating the task list. finally iterating over it to queue Creeps...

**Comment by erik-smit** [comment]
> you could test it on the ptr... 
> http://support.screeps.com/hc/en-us/articles/205999532-Public-Test-Realm

Oh, right. Forgot about that. Good idea.

>  ok this is now going even beyond the current idea,

Just a teensy bit. ;)

> return check required creeps... you get the idea...

no, i'm sorry.
this is getting too complicated for me. :D

**Comment by cyberblast** [comment]
```
task.exploit.js: 
{
  requiresCreep: () => { 
    for each source{
      if source has no miner {
        return true;
      }
    }
    if room.haulers == 0 return true;
    return false;
  }; 
[...]
}
```

but well we could also directly order them instead...  

```
task.exploit.js: 
{
  checkForRequireCreeps: () => { 
    let requiredCreeps = [];
    for each source{
      if source has no miner {
        requiredCreeps.push(new miner);
      }
    }
    if room.haulers == 0 
      requiredCreeps.push(new miner);
    for each creep in requiredCreeps{
      let room = Room.bestSpawnRoom(task);
      room.queue.push(creep);
    }
  }; 
[...]
}
```

**Comment by erik-smit** [comment]
`if source has no miner {`

How to decide if a source has no miner?
Currently there is the targetOf logic, but this only counts existing creeps.
I'm not 100% sure how this targetOf logic works.
Maybe to extend this for queued creeps?
Looking at the destiny of queued creeps?

Same for `if room.haulers == 0`.

**Comment by cyberblast** [comment]
The targetOf magic is in population.js. 
in the beginning of each loop it's iterating the creeps and creates some directories. 
1. a global directory of all creeps. Its contained in properties of the population module itself. 
e.g. call Population.typeCount['melee'] for a sum of all melee creeps. 
2. a directory per room. (room.population.typeCount)
3. targetOf directories. It reads the current target for each creep and add the creep to the targets targetOf directory. 

For spawning and queued creeps each target could get a new directory or we can use the existing targetOf. tbd.

**Comment by erik-smit** [comment]
> It reads the current target for each creep and add the creep to the targets targetOf directory.

Is this the registerAction stuff?

Ahh, population has a loop with a register function:

```
                let action = ( entry.actionName && Creep.action[entry.actionName] ) ? Creep.action[entry.actionName] : null;
                let target = action && entry.targetId ? Game.getObjectById(entry.targetId) || Game.spawns[entry.targetId] || Game.flags[entry.targetId] : null;
                if( action && target ) this.registerAction( creep, action, target, entry );
```

And:

```
_.forEach(Memory.population, register);
```

Ok, so I think I understand how targetOf appears.

Ok, so hmm... `if source has no miner {`

sources have dedicated miners, so their targetOf remains constantly set for them.
But flags are only targetOf a creep when it's travelling towards it, right?

I've just setup a room in PTR for testing this. https://screeps.com/ptr/#!/room/E68S41
I see a new room needs to be build up, so it's a bit of a hassle, but thresholds for stuff is lower, so it should be faster.

I'm thinking about what would be first/easiest to implement this for. Probably miners.

**Comment by erik-smit** [comment]
FYI, either the PTR sucks or I suck.
It's regularly superslow, I'm sometimes getting 'Circular reference to XXX' errors which unexplainably clear up.
I suck at understanding Javascript modules but I'm not sure if they related to my attempt at making a new Javascript module or not: 

```
[9:25:56 AM]Error: Circular reference to module 'creep.setup.ranger'
    at Function.extend (creep:53:21)
    at Object.module.exports.loop (main:8:11)
    at __mainLoop:1:52
```

Could this be caused by: https://github.com/cyberblast/screeps.ocs.internal/commit/8d7ce4cc7be34664cfc16e936d92ac1ae62e3148 
I've just added these two files and don't reference them anywhere.

**Comment by erik-smit** [comment]
```
E68S41 > Good morning miner-650-1!
[9:40:04 AM]E68S41 > Good morning miner-650-2!
[9:40:04 AM]E68S41 > Good morning worker-2000-1!
[9:40:54 AM]E68S41 > Good night miner-650-1!
[9:40:54 AM]E68S41 > Good night miner-650-2!
[9:40:54 AM]E68S41 > Good night worker-2000-1!
[9:46:18 AM]Corrupt creep without population entry!! : miner-650-1
[9:46:18 AM]TypeError: Cannot read property 'flee' of undefined
    at get (creep:330:34)
    at Creep.run (creep:145:25)
    at run (creep:59:38)
    at baseForOwn (/opt/ptr/engine/node_modules/lodash/index.js:2046:14)
    at Function.<anonymous> (/opt/ptr/engine/node_modules/lodash/index.js:3346:13)
    at Function.Creep.loop (creep:60:15)
    at Object.module.exports.loop (main:22:11)
[9:46:22 AM]Corrupt creep without population entry!! : miner-650-2
[9:46:22 AM]TypeError: Cannot read property 'flee' of undefined
    at get (creep:330:34)
    at Creep.run (creep:145:25)
    at run (creep:59:38)
    at baseForOwn (/opt/ptr/engine/node_modules/lodash/index.js:2046:14)
    at Function.<anonymous> (/opt/ptr/engine/node_modules/lodash/index.js:3346:13)
    at Function.Creep.loop (creep:60:15)
    at Object.module.exports.loop (main:22:11)
[9:46:26 AM]E68S41 > Good morning miner-650-3!
[9:46:26 AM]E68S41 > Good morning miner-650-4!
```

What... could this be caused by my changes in any way?

**Comment by cyberblast** [comment]
Yes the PTR sucks ^^ 
how did you spawn the 'corrupted' miners?

**Comment by erik-smit** [comment]
The 'corrupted' miners were spawned by the original code.

**Comment by erik-smit** [comment]
```
[11:46:29 AM]Script execution has been terminated: CPU bucket is empty
[11:46:33 AM]Script execution has been terminated: CPU bucket is empty
[11:46:37 AM]Error: Circular reference to module 'creep.Action'
    at Function.extend (creep:3:24)
    at Object.module.exports.loop (main:8:11)
    at __mainLoop:1:52
[11:46:40 AM]Script execution has been terminated: CPU bucket is empty
[11:46:44 AM]Error: Circular reference to module 'creep.Action'
    at Function.extend (creep:3:24)
    at Object.module.exports.loop (main:8:11)
    at __mainLoop:1:52
[11:46:48 AM]Script execution has been terminated: CPU limit reached
[11:46:51 AM]Script execution has been terminated: CPU bucket is empty
[11:46:55 AM]Script execution has been terminated: CPU limit reached
[11:46:59 AM]Script execution has been terminated: CPU bucket is empty
[11:47:03 AM]Script execution has been terminated: CPU bucket is empty
[11:47:11 AM]E68S41 > Good night worker-2000-1!
[11:47:11 AM]Script execution has been terminated: CPU limit reached
[11:47:16 AM]Script execution has been terminated: CPU bucket is empty
[11:47:20 AM]Script execution has been terminated: CPU bucket is empty
[11:47:25 AM]Script execution has been terminated: CPU bucket is empty
```

Should I just setup a private server for testing?

**Comment by erik-smit** [comment]
I forgot about the Public Test Realm limitting your CPU to 10, unless you unlock it for 24 hours at https://screeps.com/ptr/#!/order.

I think/hope this limit is why I'm running into my issues.

**Comment by cyberblast** [comment]
ah - yes! that makes sense. So you may need to leave some of your rooms on ptr ;)

**Comment by erik-smit** [comment]
I pushed a very bad attempt.

Some things I run into:

  * let vs. var, what to do
  * _.forEach(thing vs. thing.forEach
  * I don't really understand this javascript module/object/constructor/etc. business. Not sure what to define for 'task', so for now I just put it in creep.setup.miner to test it.
  * determining the name of a creep. We can't just look at Game.creeps, have to take the queued creeps into account also
  * determining the parts in relationship to RCL. Replicate or use the fixed/multi/etc. logic from Creep.setup?
  * integrating destiny into the behaviour-code of the roles.
  * 'setup', 'type', 'role', 'behaviour'? how to name the thing? calling it 'setup' in the spawnQueue sounds wrong.

**Comment by erik-smit** [comment]
I moved the checkRequiredCreep into task.exploit.js.

I'm not sure how to decide whether or not to spawn another miner.
Currently I use the targetOf.
population.js:register fills targetOf by registerAction if the Creep has an action with a target.

But a spawning creep does not get an action assigned in creep.js:
```
        Creep.prototype.run = function(behaviour){
            if( !this.spawning ){
```

So.... options.

  * Don't skip spawning creeps in Creep.prototype.run.
  * Extend population.js to also fill targetOf with the destiny of a creep
  * Extended spawning logic to also or only look at destiny of creeps? (Ignore targetOf, just look at destinies)

What you say?

**Comment by erik-smit** [comment]
I've changed the spawning logic to just look at the destiny of existing Creeps.
I think this is the way to go.

**Comment by erik-smit** [comment]
I'm getting some doubles with my code. Unsure why.

Also, I'm not sure if using the source as destiny is a great idea.
Driving to the source only works if the room is visible.
Maybe better to iterate over the flags and then use the regular determinatedTarget logic once in the room.
This does mean the spawning logic loops over 2 things. Sources for the spawnrooms and flags for the alternate rooms.

**Comment by erik-smit** [comment]
I think I fixed my double issues.

@cyberblast, any comments by now?

**Comment by cyberblast** [comment]
Hi, 
sorry for the delay. 
Sounds good so far :) 

About the things you've been running into: 
  * let vs. var, what to do

Always choose `let`. `var` will continue to live beyond closures. 
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/let 

  * _.forEach(thing vs. thing.forEach 

_.forEach can iterate collections, Array.forEach can only iterate arrays. 
https://lodash.com/docs/3.10.1#forEach 
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach 

  * determining the name of a creep. We can't just look at Game.creeps, have to take the queued creeps into account also 

another option would be not to define any name while queueing, but auto generate it during spawning...

  * determining the parts in relationship to RCL. Replicate or use the fixed/multi/etc. logic from Creep.setup? 

hmm, unsure about that. 
I think the main question here is what we want it to become. 
It would be possible to completely replace the setup files with tasks. If we want to go this way you should avoid using setup in combination with task. 

  * integrating destiny into the behaviour-code of the roles. creep.behaviour.miner.js uses  determinatedTarget. Rename destiny to determinatedTarget? Add logic to creep.behaviour.miner.js to take creep.data.destiny as determinatedTarget?

no, I think destiny can be something completely different for spawning something else. We could also go one step further here and merge destiny with setup into an object. like 
`{
  behaviour: miner, 
  task: exploit, 
  targetRoom: N0E0, 
  determinatedTarget: 123456789
}`
an other task could have other additional properties...

  * 'setup', 'type', 'role', 'behaviour'? how to name the thing? calling it 'setup' in the spawnQueue sounds wrong. 

I think destiny and destiny.behaviour is nice..

  * I see creep.behaviour.miner.js registers Creep.action.harvesting as an action if( !creep.action ) Population.registerAction(creep, Creep.action.harvesting, source);.
But do I understand correctly that creep.action.harvesting.js is not used by creep.behaviour.miner.js?
Because it doesn't use 'creep.action.step' in its run? 

Uhm .. yes. its just for census, as the miner is actually harvesting, but  - yes its quite dirty... 

  * I've changed the spawning logic to just look at the destiny of existing Creeps.
I think this is the way to go. 

Sounds good :)

**Comment by erik-smit** [comment]
> sorry for the delay.

No worries. :)

> another option would be not to define any name while queueing, but auto generate it during spawning...

Ahh! I like this!
This also solves the problem of having duplicate names when manually pushing creeps on the spawnQueue on the commandline.

https://github.com/cyberblast/screeps.ocs.internal/commit/2f6cc85a26ecf9489194afcfc4d7d87998204bdd ?

>I think the main question here is what we want it to become.
> It would be possible to completely replace the setup files with tasks. If we want to go this way you should avoid using setup in combination with task.

I think this should replace the current maxCount/per-spawn-thinking and replaces it with task-based thinking to work more as a colony instead of as independent spawns.

> no, I think destiny can be something completely different for spawning something else. We could also go one step further here and merge destiny with setup into an object. like

Ok, so the behaviour/action-code will use the destiny?

----

I'm not entirely sure where to go from here.

Should I try to rebuild the existing claimer, hauler, hauler, melee, miner, mineralMiner, pioneer, privateer, ranger, upgrader, warrior, worker roles into the new task-based structure?
Or should I try to implement the remote miner/hauler in the new task-based structure?
Or something else?

**Comment by cyberblast** [comment]
I think you should try to make small steps. And always try to reach a point where you can merge to dev. That is when it is running. 
You started with the exploit task. maybe there are some easier. like "guard a yellow flag". This one should quite straight forward ;) and you have an end to end PoC. 

> Ok, so the behaviour/action-code will use the destiny? 

(1) We need a mechanism where every new-spawn gets handled. We have such a point in population. There is a register method. and this one: 
```
                if( creep.spawning ) { // count spawning time
                    entry.spawningTime++;
                }
                else if( creep.ticksToLive == 1499 ){ // spawning complete
                    spawnsToProbe.push(entry.motherSpawn);
                }
```

this could be the moment where we need to evaluate the destiny. 

(2) An other approach could be to let the creep do it itself. Like here: 
```
        Creep.prototype.run = function(behaviour){
            if( !this.spawning ){
                if(!behaviour && this.data && this.data.creepType) {
                    behaviour = Creep.behaviour[this.data.creepType];
                }
                if( behaviour ) behaviour.run(this);
```
there could be a check for the creeps destiny - this would be each loop before the creeps behaviour is triggered. 

(3) Or it could be within the behaviour itself. 

I'd prefer the first option most and the third the least, but it could become required to step deeper...

**Comment by erik-smit** [comment]
> You started with the exploit task. maybe there are some easier. like "guard a yellow flag". This one should quite straight forward ;) and you have an end to end PoC.

https://github.com/cyberblast/screeps.ocs.internal/commit/2c20489116312ea2ae30e31ba7b46bd62aeb8fad ?

**Comment by erik-smit** [comment]
I was thinking about scaling the creep.
Should we base it on the rooms controller level? Like the (max)multi stuff?
Or on the amount of energy available?
What should the createCreepByQueue do with creeps that are larger than the current energy available?
Wait until enough enery is available?
Currently it just failes with ERR_NO_RESOURCES.

**Comment by cyberblast** [comment]
Hmm, yes in general we should wait until there are enough resources available. And the low queue should only be executed when the high is empty. maybe a "common" spawn probe in between both - or after both, maybe only when both are empty - unsure about that. 

Regarding the scaling... 
I think it shouldn't depend on the energy available but on the max energy capacity possible, which depends on RCL. Storing into empty extensions is the uppermost priority, so this shouldn't take long. 
This is where room selection becomes important as well. Do we want the closest room, or the highest RCL, a mixture of both? Other tasks may want to utilize several rooms/spawns. But "closest" is surely sufficient for now. 

Looking good so far! :) 
Maybe one request. I would turn around the "trigger". The "event" should come from an analysis of the environment (here: there is a defense flag) 
1. Flag Iteration each loop. 
2. if( flag is defenseFlag ) task.defense.checkForRequiredCreeps(flag)

**Comment by erik-smit** [comment]
> Hmm, yes in general we should wait until there are enough resources available

https://github.com/cyberblast/screeps.ocs.internal/commit/f5c66d1be4237054eb5663852ee45e6785d0f605 ?

**Comment by erik-smit** [comment]
> Maybe one request. I would turn around the "trigger". The "event" should come from an analysis of the 
> environment (here: there is a defense flag)
> 
> Flag Iteration each loop.
> if( flag is defenseFlag ) task.defense.checkForRequiredCreeps(flag)

Should the loop also check the current creeps if the flag is already filled?
Because then it's not really checkForRequiredCreeps but spawnCreep.

It kinda feels wrong.

**Comment by erik-smit** [comment]
> Regarding the scaling...
> I think it shouldn't depend on the energy available but on the max energy capacity possible, which depends on RCL. Storing into empty extensions is the uppermost priority, so this shouldn't take long.

Agreed.

Should I just copy the fixed/multi-logic or do you have a better idea?

> This is where room selection becomes important as well. Do we want the closest room, or the highest 
RCL, a mixture of both? Other tasks may want to utilize several rooms/spawns. But "closest" is surely suficient for now.

Another parameter could be, 'the closest non-busy spawn/room'.
With maybe a calculation depending on the distance between the rooms and the destination versus 

------

Also, I notice that Spawn.prototype.create() speaks of 'body' where createCreepByQueue speaks of 'parts'.
Would it make sense for me to create a pull request to align this?
And if so, since the documentation for createCreep() (http://support.screeps.com/hc/en-us/articles/205990342-StructureSpawn#createCreep) speaks of body, probably changing parms.parts to parms.body would be best?

**Comment by cyberblast** [comment]
> Another parameter could be, 'the closest non-busy spawn/room'.
With maybe a calculation depending on the distance between the rooms and the destination versus

Yes, maybe. 
Or we return an array of owned rooms, ordered by a custom formula, like "RCL - creeps in queue - distance - (availableEnergy/energyCapacity)" (highest first). We could then push as many creeps to the first x rooms as required until saturated at once (depending on the task). 

> changing parms.parts to parms.body would be best? 

sounds good :slightly_smiling_face:

**Comment by TafThorne** [comment]
What if the privateers became the remote haulers by collecting dropped energy or picking it of friendly bins in places without a spawn owned by the player?  That would leave all of their low level behavior the same and if they ran out of anything else to do they would (harmlessly?) start helping with the mining.  

You could the create a remote miner or prospector.  Say one per-spawn in remote rooms as a rough idea.  It would have to reach remotes so it may want more move parts than the basic in house mining team but could otherwise be the same.  When one dies a replacement could be spawned and sent out.  The privateer team already in the area could carry on moving energy and start doing a bit of mining if needs be.

**Comment by cyberblast** [comment]
yes that would surely be possible. but how many privateers do you need then? Wouldn't it be better (economical) to reduce costs by using haulers instead of fully functional privateers? 
The approach to have both in place (privateers + "specialized creeps") may be possible for a limited transition phase. But in the end I think we should have one clean and easy solution for remote mining. Because specialists and allrounders will disturb each other. e.g. Miners are build to completely clear a source in time. everything else harvesting on the same source is expensive and ineffective surplus.

**Comment by TafThorne** [comment]
I have tried out a basic _prospector_ in my code.  It does get rather upset when the privateers turn up and start mining too.  Most of the time they just pick up the energy from the ground and walk off though.  

Right now it is not providing a big improvement on income over just letting the privateers do their bit, though it is much faster initially.  If I got them to take from containers it might help a little.  

> yes that would surely be possible. but how many privateers do you need then? Wouldn't it be better (economical) to reduce costs by using haulers instead of fully functional privateers?

A unit more focused on hauling goods would be helpful.  I expect one privateer per room would be enough to maintain roads and so forth.  The privateer could also take over mining while a prospector was re-spawning.  Then you would need to calculate the number of mules needed based on how many were required to shift 3k of energy back to the main room every 300 ticks.  

If you did not use a privateer in the remote room what could maintain any roads?  Unless we get some strange worker/guard/claimer hybrid that takes mined energy and does up the room, heals the odd damaged unit and attempts to remove invaders.  Probably all too complicated.

**Comment by cyberblast** [comment]
Everything correct. 
So the current idea is this roadmap: 

1. Get the task system running. With any PoC task.  
Current creep spawning is a rough estimation of how much could be required. each creep will then search for a job. Thats a bad approach for dedicated creeps (maybe at all).  
2. Create an exploit task.  
The task is mainly used to organize creep spawning for each exploit flag found.  
3. Create remoteMiner & remoteHauler behaviour.  
If required. 
4. Get it running.  
5. Get rid of or reduce privateers.  
Part of the solution could be to give haulers 1 WORK part and if they are moving on a road which needs repair, they repair it (for only 1 tick). No detour, no delay.

**Comment by erik-smit** [comment]
> Create an exploit task.
> The task is mainly used to organize creep spawning for each exploit flag found.
> Create remoteMiner & remoteHauler behaviour.
> If required.

https://github.com/cyberblast/screeps.ocs.internal/commit/d30e0e82961ce7693ea8baf0af8e25df7ed46202

Unsure whether it would be better to expand the miner behaviour to include walking to other rooms.
Or to make it a completely seperate role.
This expanding seems to work for me.

Imho the only difference between a remoteMiner and a Miner would be the amount of MOVE bodyparts to get to the destination in a reasonable timeframe.

**Comment by cyberblast** [comment]
yes, that's why I added 

> **If required.**

to 3. 😉

**Comment by TafThorne** [comment]
Now I am back form holiday I can start trying to commit some time to coding and sharing code.  Here is my forked branch where I have begun work on my idea for remote mining.  

https://github.com/TafThorne/screeps.behaviour-action-pattern/tree/prospectors_and_mules 

I have only performed the fist basic step of generating a minder per owned room and have it exploit one source in a near by room.  I need to get a better idea of how the current code handles creep flag assignments before I can work out if I should get people to set one exploit flag per "prospector" and "mule" set that should be in a room.  It would probably be easier to do this per-flag than handle per-source in rooms with flags to start with.  

Ultimately I will aim for: 

- Identify rooms with exploit flags
- for each source in said room spawn one prospector and assign
- for each source assign enough mules to carry back 3000 energy per 300 ticks 

I expect I can store pathways with lengths and only refresh them infrequently. 

If I make any huge progress I will post back here.  I will try and make logical commits that explain what I am up to though so that people can see what my thought process and coding is like.  After that I will make a request to join the alliance & possibly start helping out on the roadmap if people would like me to.

**Comment by cyberblast** [comment]
Hi TafThorne, 
great work so far! 
But I'd strongly advice to merge with our ocs internal efforts. We are also already working on that and I don't think it makes any sense to have 2 completely different streams working on the same problem. 
You mentioned your intention to join the alliance. Maybe it would make more sense to join **before** working on this topic. But in the end it's up to you, of course...

**Comment by TafThorne** [comment]
Thanks for the positive feedback @cyberblast.  I would be very pleased to join the alliance and help out with the shared code base.  I made the fork and commits yesterday mostly to capture what I had generated while away on holiday now that I am back somewhere with a PC.

**Comment by erik-smit** [comment]
I've got remoteMiner and remoteHauler s up and running.
extra remoteHaulers get spawned based on time it took for the Hauler to walk from remote container to local storage.
Stuff to do:

  - something needs to build the rooms in the exploited rooms
  - something needs to maintain the rooms in the exploited rooms
  - something with containers

**Comment by cyberblast** [comment]
> something needs to build the rooms in the exploited rooms 

you could use a privateer for this job. you can disassemble it when it's finished. 

> something needs to maintain the rooms in the exploited rooms 

I guess you mean road repairing? We've had this idea to spent a work part for the remote hauler. It should be possible to repair the road "lossles", regading movement time. Simply everytime a bit when it is moving over it... 

> something with containers 

Ain't this included in 1 (privateers setting up the exploit room)?

> maybe start spawning a replacment remoteMiner sooner based on the walktime of the miner 

Good idea but I'd propose to leave this for tuning, when the overall thing is working...

**Comment by erik-smit** [comment]
> you could use a privateer for this job. you can disassemble it when it's finished.
> I guess you mean repairing?
> Ain't this included in 1 (privateers setting up the exploit room)?

How do you foresee this?
Let one of the creeps spawn a privateer with destiny when it FIND_CONSTRUCTION_SITES?
Extend existing task?
Create new task.privateer?

> Good idea but I'd propose to leave this for tuning, when the overall thing is working...

Agree. I was just listing random TODO stuff

**Comment by cyberblast** [comment]
hmm. or like when the flag is found the first time you calc required paths (like home storage -> remote source) and check each pos if there is a road.

**Comment by TafThorne** [comment]
Hey @erik-smit how was https://github.com/erik-smit/screeps.behaviour-action-pattern/commits/issue-206 looking?  

I think rather than aiming for perfection getting _some_ improvement would be OK.  Sending one miner & pioneer per source would be a start.  Then if there are containers in the room we could maybe use those to judge how many haulers are needed.  

Everytime (maybe do a check every 600 ticks or so) we see energy on the ground we up the target number of haulers by +1 if we have the current target number of haulers going.  If we see empty containers in a room we decrement the target hauler number by one (or set it to the current number of haulers if it is a lot).  I had something like that in place for my in-room miners before joining up and it kinda worked.  

I have no idea how hard that sort of calculation would be to do in this code base but I'll try and give it look later today.

**Comment by TafThorne** [comment]
Silent closure with no mention of a patch?  Does this mean it is dropped or that it went into a code line without mentioning this ticket?

**Comment by whardeman** [comment]
@TafThorne my mistake not mentioning the recent PR
